### ISSUE TITLE
Build rpm package for Firefox

### ISSUE THREAD

**Issue by SilvanScherrer** [bitwiseworks/mozilla-os2#142]

**Comment by dmik** [comment]
This is almost done, however I had to fix quite a bit of stuff in various RPMs (autoconf213, autoconf, texinfo, nspr, libc os2tk45). Note also that we will use the existing firefox.exe code (nsBrowserApp.cpp) as a /usr/bin launcher (instead of run! and similar) as this is what it effectively is. The only essential difference will be that the original firefox.exe loads XUL.DLL from its own dir, while the /usr/bin one will load from ../lib/fireflx-XYZ. This has a number benefits:

- No separate process will be started (saves several hundred KB of mem and some other resources).
- Stroked icon appearance will be preserved w/o extra waiting process.
- No separate code, just a few modifications to the original launcher.

**Comment by dmik** [comment]
Sad to admit but `make install` for Firefox fails in the "check for duplicates" python code which freezes the FS so that even C-A-D doesn't work. Too sad as it means no FF 45 release until it's fixed. And who knows how much time it will require to find the problem and fix it. It may involve many apps and many libs as usual.

**Comment by dmik** [comment]
I have a gut area feeling that it's due to the shitty SMP support in the kernel (I have 4 cores running now and this appears to be too much for our SMP kernel).

**Comment by dmik** [comment]
No, it doesn't seem to have anything to do with SMP safety. At least booting in /MAXCPU=1 mode as well as trying MARKEXE MPUNSAFE on python doesn't help. A deep python debug is necessary.

**Comment by dspiatkowski** [comment]
Is this honestly the only way to go???

Even in the Winxx world one does not simply dump every app into a common
directory...I just can't for the life of me understand why we would want to
force every user out there to go to \usr\bin to run Firefox...please tell me
I'm wrong...in BIG BOLD CAPS...but I would suspect that this is a non-starter
for many!

Regards,
-Dariusz

**Comment by dryeo** [comment]
On 05/15/17 02:35 PM, Dmitriy Kuminov wrote:
> This has a number benefits:
>
>   * No separate process will be started (saves several hundred KB of mem
>     and some other resources).
>   * Stroked icon appearance will be preserved w/o extra waiting process.
>   * No separate code, just a few modifications to the original launcher.

I don't understand. Previously, if only running Firefox, you could just 
run firefox.exe and if also running Thunderbird you needed run!. Now if 
you want to run Firefox and Thunderbird, the user is going to have to 
manually install run! into @UNIXROOT/usr/bin, unless you make an RPM for 
Firefox!g.exe for users who want to run multiple mozilla apps. Adding K 
to the run! command has always been optional if you wanted to waste 
resources and have a stroked icon. A cmd file wrapper or sh file (with 
the firefox icon associated with it) wrapper like all the unix type OSes 
uses is another option if a user wants a stroked icon.

**Comment by SilvanScherrer** [comment]
@dryeo we will of course take that into account and use LIBPATHSTRICT for the new/adapted launcher. As else it would be a really bad idea.

**Comment by SilvanScherrer** [comment]
@dspiatkowski you can always use the zip to install it. We will provide the zip and the rpm, like we do with all our rpm.

**Comment by dmik** [comment]
@dspiatkowski I'm not aware of your problem. What's wrong with having a launcher on PATH? It just gives benefits and no downsides. 

@dryeo of course the launcher will set BEGINLIBPATH and LIBPATHSTRICT. In fact, this should have been done long ago and not run! or cmd wrapper would ever be needed. Anyway, from the user's perspective nothing will change except that extra memory will be saved for their convenience.

**Comment by dspiatkowski** [comment]
Hi Dmitriy,

On Wed, 17 May 2017 02:12:01 -0700, Dmitriy Kuminov wrote:

>@dspiatkowski I'm not aware of your problem. What's wrong with having a launcher on PATH? It just gives benefits and no downsides. 

So when I read a statement that says (in summary) that we are going to use an
RPM package for Firefox (the application) it implies that I am handing over the
control of the install to the ANPM or the YUM/RPM cli interface. As such, I no
longer have the say over where on the disk the stuff is actually put, meaning,
in my case all ANPM installs go to \usr\... location.

To me that reads as follows: FF, when installed as a RPM package will be
installed somewhere in the \usr\... tree.

Now your response above suggests that having launcher in the PATH is the only
impact. So are we saying the RPM package will allow me to install somewhere
other then \usr\... but a launcher only will be tossed into \usr\bin?

To summarize, I had a well thought out structure in place to manage my DLLs
prior to moving to RPM. I did move to RPM so that the 'system' type stuff (that
would include small utilities, etc) would have more consistency from an
install/maintenance perspective, and certainly if that meant I could be more
productive in terms of testing/debugging assistance that was a +ve outcome. 

But now it appears actual functional applications are also going to be put into
the \usr\... tree, which I simply think is a bad idea. If that is the
direction, what is the benefit to this approach?

Did I missunderstand that RPM install of Firefox statement? Because going back
to a ZIP install simply defeats the whole purpose of standardizing on RPM
installs.

Thanks,
-Dariusz

**Comment by wztest** [comment]
> "Too sad as it means no FF 45 release until it's fixed."
It was a long journey with all fixes and tools in place to get this build up and running and now we are waiting because of RPM issues? Why not put a zip distribution out for users which prefer it and can run and test it already?

**Comment by dmik** [comment]
We have a plan to make an intermediate ZIP-only release for testing purposes. It won't need any special launcher.

**Comment by wztest** [comment]
That sound's good ! Any news about printing?

**Comment by dmik** [comment]
@wztest nothing new about printing so far.

**Comment by dmik** [comment]
The python freeze issue was fixed via LIBCx (see https://github.com/bitwiseworks/libcx/issues/36 for details and especially https://github.com/bitwiseworks/libcx/commit/de89d077049e5f571597dfb99ffa9ac7c1fa231f). Now Firefox RPM gets successfully built.

The only thing left is aligning firefox.exe so that it could search for XUL.DLL in `/@unixroot/usr/lib/firefox-X.Y.Z` besides the current directory. Should be minor.

**Comment by dmik** [comment]
The launcher is done; it's now only a question of doing an RPM rebuild to get it all packed. But first we will do an update to the latest ESR 45 (which is 45.9.0). Created #217 for that.

**Comment by dmik** [comment]
Actually the first RPM package has been already released to ArcaOS exp repo on Aug, 31. In general, It works very well except a side problem mentioned in http://trac.netlabs.org/rpm/ticket/269. This problem will be fixed in the next RPM which will have a GA status (to be released in a few days).

**Comment by LewisR** [comment]
To be clear, it's the arcanoae-exp repo, and the GA will be in arcanoae-sub (not arcanoae-rel). I guess we should come up with an arcanoae-sub-exp repo instead of a public one for testing reserved features. (New repos coming soon.)

zips will always be available at Netlabs, but RPMs are reserved features for Drivers & Software and ArcaOS Support & Maintenance subscribers.

**Comment by dmik** [comment]
@LewisR Yes, if exp is open than sub-exp is needed. Though I think that `exp` should be closed for general public as it raises too many problems (we're planning to do so at bww). So there will be an `exp->stage` flow directed by devs and then either `stage->res` or `stage->sub` both directed by release managers. But that's a bit off-topic here.

**Comment by dmik** [comment]
firefox-45.9.0-2 RPMs are built and uploaded to ArcaNoae repositories.
